rust/noosphere-ns/src/records.rs

```diff
@@ -0,0 +1,320 @@
+use anyhow::{anyhow, Error};
+use cid::Cid;
+use serde::{
+    de::{self, Deserializer, MapAccess, SeqAccess, Visitor},
+    ser::{SerializeStruct, Serializer},
+    Deserialize, Serialize,
+};
+
+use std::{
+    fmt,
+    str::FromStr,
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+
+/// An [NSRecord] is a struct representing a record stored in the
+/// Noosphere Name System's DHT containing a [Cid] of the
+/// result, as well as a TTL expiration as seconds from Unix epoch.
+///
+/// # Serialization
+///
+/// When transmitting records across the network, they're encoded
+/// as JSON-formatted UTF-8 strings. The [NSRecord::to_bytes]
+/// and [NSRecord::from_bytes] methods respectively handle the
+/// serialization and deserialization in this format.
+///
+/// Fields are mapped to the corresponding properties and JSON types:
+///
+/// * `cid` => `"cid" as String`
+/// * `expires` => `"exp" as Number`
+///
+/// An example of the serialized payload structure and
+/// conversion looks like:
+///  
+/// ```
+/// use noosphere_ns::{Cid, NSRecord};
+/// use std::str::FromStr;
+///
+/// let cid_str = "bafkreibme22gw2h7y2h7tg2fhqotaqjucnbc24deqo72b6mkl2egezxhvy";
+/// let expires = 1667262626;
+///
+/// let record = NSRecord::new(Cid::from_str(cid_str).unwrap(), expires);
+/// assert_eq!(record.cid.to_string(), cid_str);
+/// assert_eq!(record.expires, expires);
+///
+/// let bytes = record.to_bytes().unwrap();
+/// let record_str = "{\"cid\":\"bafkreibme22gw2h7y2h7tg2fhqotaqjucnbc24deqo72b6mkl2egezxhvy\",\"exp\":1667262626}";
+/// assert_eq!(&String::from_utf8(bytes.clone()).unwrap(), record_str);
+/// assert_eq!(NSRecord::from_bytes(bytes).unwrap(), record);
+/// ```
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub struct NSRecord {
+    /// The link to the resolved sphere's content.
+    pub cid: Cid,
+    /// TTL expiration time as seconds from Unix epoch.
+    pub expires: u64,
+}
+
+impl NSRecord {
+    /// Creates a new [NSRecord].
+    pub fn new(cid: Cid, expires: u64) -> Self {
+        Self { cid, expires }
+    }
+
+    /// Creates a new [NSRecord] with an expiration `ttl` seconds from now.
+    pub fn new_from_ttl(cid: Cid, ttl: u64) -> Result<Self, Error> {
+        let expires: u64 = SystemTime::now()
+            .checked_add(Duration::new(ttl, 0))
+            .ok_or_else(|| anyhow!("Duration overflow."))?
+            .duration_since(UNIX_EPOCH)
+            .map_err(|e| anyhow!(e.to_string()))?
+            .as_secs();
+        Ok(Self { cid, expires })
+    }
+
+    /// Creates a new [NSRecord] from serialized bytes. See [NSRecord]
+    /// for serialization details.
+    pub fn from_bytes(bytes: Vec<u8>) -> Result<Self, Error> {
+        let string = String::from_utf8(bytes).map_err(|e| anyhow!(e.to_string()))?;
+        serde_json::from_str(&string).map_err(|e| anyhow!(e.to_string()))
```

Comment:
No opinion I think on the JSON representation, just the quick and easy solution for now. If we do not need to "staple" NS data onto the record, this could simply be utf8 encoded Ucan

Reply:
Yah, I think even if JSON I would vote for using something ready to go out of https://github.com/subconsciousnetwork/noosphere/blob/main/rust/noosphere-storage/src/encoding.rs

I would prefer `CBOR`, but `DagJsonCodec` should also work with those methods (caveat we haven't used it anywhere).